### PR TITLE
fix(schemas): ensure schema permission 

### LINKED ISSUE #627: Fix `ensure_valid_schema_grant()` to check for revocations
Fix `ensure_valid_schema_grant()` to check for revocations at current block:

- [ ] Delegation has been revoked between provider and delegator
- [ ] Schema permissions were revoked for delegation 
- [ ] Specific permission for a single schema id was revoked

except for block zero.

### PR DESCRIPTION
**Pull request by rlaferla**
Ensure that schema grant validation works.
A schema permission is valid for a block-number if:
- a delegation relationship has not been revoked.
- a schema permission has not been revoked.

If a schema is revoked at block-number X:
- the schema is valid for the block-number X
  and prior.
- the schema is invalid after block-number X.

Add benchmarking helper to messages-pallet.
Adding a benchmarking helper facilitated the
removal of duplicating functionality for
benchmarks.

Update benchmarks.

issue-627

### CODE REVIEW

**Review by wilwade — CHANGES_REQUESTED** [review]
I'm surprised it didn't break any tests, but I think we have some tests to add.

**Comment by enddynayn** [comment]
Yeah I think it did not break any test because some are mocked.

File: pallets/msa/src/lib.rs [hunk 1]
```diff
@@ -1014,6 +1016,11 @@ impl<T: Config> Pallet<T> {
 		)
 	}
 
+	// |-----|-----|------|------|------|-
```
Comment (enddynayn) [inline pallets/msa/src/lib.rs:1019]: remove

File: pallets/msa/src/tests.rs [hunk 2]
```diff
@@ -2566,3 +2531,123 @@ fn schema_permissions_trait_impl_try_get_mut_schema_success() {
...
+	new_test_ext().execute_with(|| {
+		set_schema_count::<Test>(2);
+
+		// create delegation relationship
+		let provider = Provider(1);
+		let delegator = Delegator(2);
+		let schema_grants = vec![1, 2];
+		assert_ok!(Msa::add_provider(provider, delegator, schema_grants));
+
+		// create delegation relationship
```
Comment (enddynayn) [inline pallets/msa/src/tests.rs:2586]: move comment

File: pallets/msa/src/lib.rs [hunk 3]
```diff
@@ -1371,7 +1389,7 @@ impl<T: Config> DelegationValidator for Pallet<T> {
 	}
 }
 
-impl<T: Config> SchemaGrantValidator for Pallet<T> {
+impl<T: Config> SchemaGrantValidator<T::BlockNumber> for Pallet<T> {
 	/// Check if provider is allowed to publish for a given schema_id for a given delegator
 	///
 	/// # Errors
```
Comment (wilwade) [inline pallets/msa/src/lib.rs:1395]: We should bubble up the new error here as well.

File: pallets/msa/src/lib.rs [hunk 4]
```diff
@@ -1400,6 +1419,7 @@ impl<T: Config> SchemaGrantValidator for Pallet<T> {
 		provider: Provider,
 		delegator: Delegator,
 		_schema_id: SchemaId,
+		block_number: T::BlockNumber,
```
Comment (wilwade) [inline pallets/msa/src/lib.rs:1422]:
```suggestion
		_block_number: T::BlockNumber,
```

**Review by wilwade — COMMENTED** [review]
Few notes. Feel free to ignore as it is in draft and you might have already fixed them. I'm just catching up to understand the upcoming changes.

File: pallets/msa/src/lib.rs (hunk 3 above)
Comment (enddynayn) [reply pallets/msa/src/lib.rs:1400]: looks like `ensure_valid_delegation` this one too.

File: pallets/msa/src/lib.rs [hunk 5]
```diff
@@ -1055,7 +1057,10 @@ impl<T: Config> Pallet<T> {
 		let current_block = frame_system::Pallet::<T>::block_number();
 		let requested_block = match block_number {
 			Some(block_number) => {
-				ensure!(current_block >= block_number, Error::<T>::DelegationNotFound);
+				ensure!(
+					current_block >= block_number,
+					Error::<T>::CannotPredictValidityPastCurrentBlock
```
Comment (shannonwells) [inline pallets/msa/src/lib.rs:1062]: This should go in the errors list in the comments as well

File: pallets/msa/src/tests.rs [hunk 6]
```diff
@@ -1464,7 +1467,7 @@ fn signed_extension_revoke_delegation_by_provider_fails_when_no_delegation() {
 		let (_, provider_pair) = create_account();
 		let provider_account = provider_pair.public();
 		let (delegator_msa, delegator_pair) = create_account();
-		let delegator_account = delegator_pair.public();
+		let _delegator_account = delegator_pair.public();
```
Comment (shannonwells) [inline pallets/msa/src/tests.rs:1470]: why not just delete `_delegator_account`?

File: pallets/msa/src/tests.rs [hunk 7]
```diff
@@ -1451,7 +1454,7 @@ fn signed_extension_revoke_delegation_by_provider_fails_when_no_provider_msa() {
 		let (provider_pair, _) = sr25519::Pair::generate();
 		let provider_account = provider_pair.public();
 
-		let (delegator_msa, delegator_pair) = create_account();
+		let (delegator_msa, _delegator_pair) = create_account();
```
Comment (shannonwells) [inline pallets/msa/src/tests.rs:1457]: why not just make this `_`  ?

**Review by shannonwells — APPROVED** [review]
Looks good to me aside from two non-blocking nitpicks.

**Review by wilwade — APPROVED** [review]

File: pallets/msa/src/benchmarking.rs [hunk 8]
```diff
@@ -192,17 +192,17 @@ benchmarks! {
...
-		// https://substrate.stackexchange.com/a/4430/2060
-		let m in 1 .. 3_000;
-		for j in 0 .. m {
-			let mortality = 49;
-			register_signature::<T>(mortality as u32);
-		}
-	}: {
-		Msa::<T>::on_initialize(200u32.into());
-	}
+	// on_initialize {
```
Comment (enddynayn) [inline pallets/msa/src/benchmarking.rs:195]: commented out for quicker test runs.

File: common/primitives/src/messages.rs [hunk 9]
```diff
@@ -10,6 +10,15 @@ use sp_std::{prelude::*, vec};
 #[cfg(feature = "std")]
 use utils::*;
 
+#[cfg(feature = "runtime-benchmarks")]
```
Comment (aramikm) [inline common/primitives/src/messages.rs:13]: I would create a new rs file and put all these logic inside it and loads it only under `feature = "runtime-benchmarks"` in lib.rs

like 
```
#[cfg(feature = "runtime-benchmarks")]
pub mod benchmarking_helper;
```

File: common/primitives/src/messages.rs [hunk 10]
```diff
@@ -130,6 +139,43 @@ impl<T> BlockPaginationResponse<T> {
 	}
 }
 
+#[cfg(feature = "runtime-benchmarks")]
+/// A trait for helping setup state for running benchmarks.
+pub trait BenchmarkHelper<AccountId> {
```
Comment (shannonwells) [inline common/primitives/src/messages.rs:144]: I would love to see more explanation here about what this helper does and how to use it with benchmarks.  Maybe also link to the StackOverflow/StackExchange question.

**Review by shannonwells — APPROVED** [review]
Wow what a pain. Looks good.

File: pallets/messages/src/benchmarking.rs [hunk 11]
```diff
@@ -1,61 +1,83 @@
+#![cfg(feature = "runtime-benchmarks")]
```
Comment (aramikm) [inline pallets/messages/src/benchmarking.rs:1]: why do you need this if you are only including it conditionally in messages/src/lib.rs
```

#[cfg(feature = "runtime-benchmarks")]
mod benchmarking;
```

File: common/primitives/src/messages.rs [hunk 12]
```diff
@@ -130,6 +139,43 @@ impl<T> BlockPaginationResponse<T> {
...
+
+	/// Sets a publickey to an MSA.
+	fn add_key(msa_id: MessageSourceId, key: AccountId) -> DispatchResult;
+
+	/// Sets the schema count.
+	fn set_schema_count(schema_id: SchemaId);
+}
+
+#[cfg(feature = "runtime-benchmarks")]
+impl<AccountId> BenchmarkHelper<AccountId> for () {
```
Comment (aramikm) [inline common/primitives/src/messages.rs:160]: why do you need this?

File: common/primitives/src/messages.rs (hunk 10 above)
Comment (rlaferla) [reply common/primitives/src/messages.rs:144]: Me too

Comment (aramikm) [inline common/primitives/src/messages.rs:144]: this helper is a bit weird since it is a mix of functionalities from two different pallets

**Review by aramikm — COMMENTED** [review]
I like the general approach 🎆 , just needs some clean up and refinement

File: pallets/messages/src/benchmarking.rs (hunk 11 above)
Comment (enddynayn) [reply pallets/messages/src/benchmarking.rs:1]: oh yeah. your right.

File: common/primitives/src/messages.rs (hunk 12 above)
Comment (enddynayn) [reply common/primitives/src/messages.rs:160]: it makes things easier
```
type Helper = ()
```

File: common/primitives/src/messages.rs (hunk 10 above)
Comment (enddynayn) [reply common/primitives/src/messages.rs:144]: I agree. I did not want to spend more time fixing the bug. We still need to remove the duplication from the schemas. So whoever does that can improve this.

File: common/primitives/src/messages.rs (hunk 12 above)
Comment (aramikm) [reply common/primitives/src/messages.rs:160]: Maybe we need a comment to describe this is mainly for tests

**Pull request merged**